lib/std/crypto/pbkdf2.zig

```diff
@@ -0,0 +1,227 @@
+//                 c          iteration count, a positive integer
+//                 dkLen      intended length in octets of the derived
+//                            key, a positive integer, at most
+//                            (2^32 - 1) * hLen
+//
+// Output:         DK         derived key, a dkLen-octet string
+
+// Based on Apple's CommonKeyDerivation, based originally on code by Damien Bergamini.
+
+pub fn pbkdf2(derivedKey: []u8, password: []const u8, salt: []const u8, rounds: u32, comptime Hash: type) void {
+    assert(rounds >= 1);
+
+    const dkLen = derivedKey.len;
+    const hLen = Hash.digest_length;
+    const Prf = crypto.auth.hmac.Hmac(Hash);
+
+    // FromSpec:
+    //
+    //   1. If dkLen > (2^32 - 1) * hLen, output "derived key too long" and
+    //      stop.
+    //
+    assert(dkLen > 0 and dkLen <= (1 << 32 - 1) * hLen);
+
+    // FromSpec:
+    //
+    //   2. Let l be the number of hLen-octet blocks in the derived key,
+    //      rounding up, and let r be the number of octets in the last
+    //      block
+    //
+    const l = (dkLen + hLen - 1) / hLen;
+    var r = dkLen % hLen;
+    r = if (r != 0) r else hLen;
+
+    // FromSpec:
+    //
+    //   3. For each block of the derived key apply the function F defined
+    //      below to the password P, the salt S, the iteration count c, and
+    //      the block index to compute the block:
+    //
+    //                T_1 = F (P, S, c, 1) ,
+    //                T_2 = F (P, S, c, 2) ,
+    //                ...
+    //                T_l = F (P, S, c, l) ,
+    //
+    //      where the function F is defined as the exclusive-or sum of the
+    //      first c iterates of the underlying pseudorandom function PRF
+    //      applied to the password P and the concatenation of the salt S
+    //      and the block index i:
+    //
+    //                F (P, S, c, i) = U_1 \xor U_2 \xor ... \xor U_c
+    //
+    //  where
+    //
+    //            U_1 = PRF (P, S || INT (i)) ,
+    //            U_2 = PRF (P, U_1) ,
+    //            ...
+    //            U_c = PRF (P, U_{c-1}) .
+    //
+    //  Here, INT (i) is a four-octet encoding of the integer i, most
+    //  significant octet first.
+    //
+    //  4. Concatenate the blocks and extract the first dkLen octets to
+    //  produce a derived key DK:
+    //
+    //            DK = T_1 || T_2 ||  ...  || T_l<0..r-1>
+
+    var prevBlock: [hLen]u8 = undefined;
+    var newBlock: [hLen]u8 = undefined;
+
+    var block: u32 = 0; // Spec limits to u32
+    while (block < l) : (block += 1) {
+
+        // U_1 = PRF (P, S || INT (i))
+        const blockIndex = mem.toBytes(mem.nativeToBig(u32, block + 1)); // Block index starts at 0001
+        var ctx = Prf.init(password);
+        ctx.update(salt);
+        ctx.update(blockIndex[0..]);
+        ctx.final(prevBlock[0..]);
+
+        // Choose portion of DK to write into (T_n) and initialize
+        const offset = block * hLen;
```

Comment:
Could this ever overflow?

Reply:
I think it's possible. I'm exploring. Since `block` is u32, and `hLen` could be anything, I assume `offset` could be inferred to be u32. If `dkLen` were larger than 2^32 bytes (which would be unthinkable, but technically legal), then I think this will overflow. Trying to test this, but the time required may be too long to verify.

I believe the solution is to define `offset` as `usize`:

        const offset: usize = block * hLen;

It's not possible for `offset` to be larger than the size of `derivedKey`, and I assume that zig limits this to `usize`, correct?

Reply:
If it's *supposed* to overflow, then you need to use the `*%` operator. Otherwise you are asserting overflow will never happen with the `*` operator (causing Undefined Behavior in unsafe build modes!). However if the correctness of the code depends on that operation not overflowing, best to use `*`. It's not entirely clear to me what the answer is here without more careful study of the code. One thing to be careful of is that `usize` can be 32 bits and even 16 bits depending on pointer size, so don't assume it's necessarily larger than a u32. If you need a certain number of bits, use that integer! You could even use a `u33` if you need to.

Reply:
This must not overflow. That would be invalid (and create security problems).

`offset` doesn't need a specific number of bits. It needs to be able to hold any offset into a slice (which is passed from the caller). Can slices be larger than `usize`?

I've been able to cause an overflow, even using `usize` on a 64-bit system. It took several hours to test, so I need to do some work to explore the issue with a faster piece of code. (PBKDF2 is intentionally slow. No one should ever use it in the way we're discussing. But this is security code and must have correct behavior over all legal inputs, even if they're absurd.)

Here is the line of code:

        const offset: usize = block * hLen;

I don't know the exact value of `block` yet, but it's a u32, and hLen is 20 (I don't know what type it is). I expect that `block * hLen` is just larger than a u32 (dkLen in this test is 2^33).

In the above line of code, does zig widen `block` and `hLen` first, or multiply first and then widen?

```
Test [7/7] test "Very large dkLen"... integer overflow
/Users/rnapier/projects/zig/lib/std/crypto/pbkdf2.zig:111:30: 0x100cb5f14 in pbkdf2 (test)
        const offset: usize = block * hLen;
                             ^
```

Reply:
u32 * u32 = u65 actually

Reply:
I think the fix might be
```zig
debug.assert(hLen <= math.maxInt(u32));
const offset: u64 = @as(u64, block) * hLen;
```

Reply:
> Can slices be larger than `usize`?

no, the type of the `len` property of slices is `usize`.

> I don't know what type it is, what type do you get for `const digest_length = 20`?)

[comptime_int](https://ziglang.org/documentation/master/#Primitive-Types). Think Python integers, but the value has to be known by the compiler.

> In the above line of code, does zig widen `block` and `hLen` first, or multiply first and then widen?

multiply first, then widen. the type of the multiplication will be the result of [peer type resolution](https://ziglang.org/documentation/master/#Peer-Type-Resolution) between block and hLen